### ISSUE TITLE
[🐛 bug report] Mitsuba 3 Test Suite Crashes with Python Error due to use of Assert Statements

### ISSUE THREAD

**Issue by Microno95** [mitsuba-renderer/mitsuba3#928]
## Summary

The python instance running the pytest test suite crashes when a test case fails due to a RuntimeError raised from C++ in `mitsuba` or `drjit`. Also happens with failing `assert` statements.

## System configuration
```
System information:

  OS: Windows-10
  CPU: AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD
  GPU: NVIDIA GeForce GTX 1080
NVIDIA GeForce RTX 3090
  Python: 3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:12:32) [MSC v.1929 64 bit (AMD64)]
  NVidia driver: 537.13
  CUDA: 11.7.99
  LLVM: 15.-1.-1

  Dr.Jit: 0.4.3
  Mitsuba: 3.4.0
     Is custom build? False
     Compiled with: MSVC 19.37.32824.0
     Variants:
        scalar_rgb
        scalar_spectral
        scalar_mono_polarized
        scalar_spectral_polarized
        cuda_rgb
        cuda_rgb_double
        cuda_spectral
        cuda_spectral_double
        cuda_ad_rgb
        cuda_ad_rgb_double
        llvm_rgb
        llvm_rgb_double
        llvm_spectral
        llvm_spectral_double
        llvm_ad_rgb
        llvm_ad_rgb_double
```

## Description

Not quite sure what exactly is happening under-the-hood, but I think it relates to the fact that, in the process of generating a traceback, python calls `repr` on certain variables that causes `drjit` to abort the process. Digging into it with LLDB, I saw one case in relation to a `repr` call on a variable that led to `drjit` to be unhappy about trying to extract the value of a placeholder variable.

In some cases, it definitely makes sense that a fatal error will occur when variables are in an inconsistent state such as when certain operations fail during loop recording and the traceback attempts to extract the values of placeholder variables (though it would be nice if this could be passed up as an error before aborting the instance), but the minimal case below should be recoverable and it certainly isn't.

## Steps to reproduce

1. Remove any of the `primal_ref.exr` images from `resources/data/tests/integrators`
2. Run `test_ad_integrators.py` test suite with `pytest` only (no exit on failure flag)
4. This will raise `RuntimeError: I/O error while attempting to open file: No such file or directory` during the test which should theoretically simply fail the test and allow the suite to move on
5. As a consequence the next test case will fail with `Fatal Python error: Aborted` - or one much later, but consistently leads to a fatal python error after the error above.

## A Solution

The interim solution is to replace assertions with `pytest.fail` calls that provide the same information, and can even be augmented with more informative failure messages.

More broadly, it might make more sense to identify exactly the string of function calls that lead to SIGABRT in the process. I've seen one case with `repr`, but it could point to a broader issue.

I'm happy to implement the interim solution above and create a PR for it.

**Comment by njroussel** [comment]
Hi @Microno95 

We've been well aware of this and never looked into it. It's one of those things where we just figure that we're better off fixing the first failing test anyway. I do agree that sometimes it can be quite annoying.

I'll run this through the debugger to see what is happening.

**Comment by njroussel** [comment]
I unfortunately don't think there is anything we can do here. 

I can't tell if it's Python or PyBind, but one the two does not want to clean up its memory after the test has failed. In fact, it will decide to do so randomly in a later test.
The issue with this is that on the beginning of each test we cleanup the JIT state, and whenever Python/PyBind finally decides to cleanup the failed test's local storage, Dr.Jit will see request to clear memory/pointers that no longer exist (as they were already cleaned up at the beginning of the current test). 

We could eventually launch tests in a clean new process every time, but that overhead is undesirably for a lot of reasons.

This might improve when we migrate to https://github.com/wjakob/nanobind.

**Comment by wjakob** [comment]
I think that many parts of the Dr.Jit python bindings are a bit flaky when it comes to reference counting in the presence of exceptions being thrown around. The upcoming nanobind-based rewrite of Dr.Jit `nanobind_v2` branch will address that to a certain extent, and other changes in Mitsuba should further help. For example, I want to make Python-based implementations of plugins (trampoline based) collectable even when they occur in reference cycles.

**Comment by wjakob** [comment]
One issue here btw is also Pytest. When an exception is raised, it walks through the stack and creates a ton of references to various things to create a nice error message at the end. So the generation of the problem description at the end of the test run indeed calls `repr` on things that were kept alive all this time, with numerous cleanup steps in between.

**Comment by Microno95** [comment]
Right, that makes a lot of sense and explains why monkey-patching `drjit/router.py` helps alleviate some of the issues since it prevents the variable from being scheduled for computation.
